### PR TITLE
Add visual representation for filters

### PR DESCRIPTION
**Pull request by jas7457**
### PR Summary: 

You can now use color swatches and thumbnails for your store's metaobject filters.

### Why are these changes introduced?

This will allow merchants using the Search & Discovery app a way to choose between a swatch or thumbnail visual style for certain filter types (to date: only metafield references to metaobjects).

### What approach did you take?
I followed the [figma design](https://www.figma.com/file/DfFkTjUypdH1oDZSBfByNt/Storefront-Filtering-enhancements?node-id=6842%3A105934&mode=dev), and tried to componentize the visual representation so it can be used outside of the context of filters as well.

When wrapped in a `.visual-display-parent` and either `.visual-display-parent--swatch` or `.visual-display-parent--thumbnail`, hover, active, disabled, and focus states can be applied as well.

### Other considerations
Ensure that this can be reused outside of the context of filters.

It's worth noting that there is a TON of unnecessary duplication between the mobile filters and the non-mobile filters. I added a few classes to each to cut down on the duplication of css rules (`.facets__label`, `.facets-layout-grid`, `.facets-layout-list`, `.facet-checkbox__text` to name a few), but I think that we should take the time to refactor filters from the ground up at a later date to get rid of all/most of the duplication of HTML and CSS.

### Decision log

| # | Decision | Alternatives | Rationale | Downsides |
|---|---|---|---|---|
| 1 | Keep visual representation generic and not only filter-aware | Have less css/code and only allow it to work inside of the context of filters | Other teams or consumers may want to show a visual representation anywhere else in the theme. | The HTML structure and CSS gets a bit bloated with this approach, but allows for better componentization and reusability. |

### Visual impact on existing themes

In the image below, I am showing the different possible states. In order:
1. Default
2. Active
3. Hover
4. Focus
5. Disabled
6. Empty

image

### Testing steps/scenarios

Testing directions are listed [here](https://github.com/Shopify/dawn-private/issues/251).

[Theme upload](https://github.com/Shopify/dawn/files/13047474/Dawn-11.0.0.zip)

### Demo links

- [Editor](https://admin.shopify.com/store/jsonwebdev/themes/157816717334/editor?previewPath=%2Fcollections%2Fall)

### Checklist
- [x] Added PR summary for [release notes](https://themes.shopify.com/themes/dawn/styles/default#ReleaseNotes)
- [x] Requested review from UX (Only for changes that are affecting the experience or perceivable visual details)
- [x] Created a ticket for the [help.shopify.com](https://help.shopify.com) documentation team about updates to theme settings. (Internal-only task)
- [x] Followed [theme code principles](https://github.com/Shopify/dawn/blob/main/.github/CONTRIBUTING.md#theme-code-principles)
- [x] Linted with [Theme Check](https://github.com/Shopify/theme-check)
- [x] Tested on [mobile](https://shopify.dev/themes/store/requirements#mobile-browser-requirements)
- [x] Tested on [multiple browsers](https://shopify.dev/themes/store/requirements#desktop-browser-requirements)
- [x] Tested for [accessibility](https://shopify.dev/themes/best-practices/accessibility)

### CODE REVIEW

File: snippets/visual-display.liquid
Comment (carolineschnapp) [inline snippets/visual-display.liquid:36]: indenting too big here

**Review by patrickracicot — COMMENTED** [review]
This is some good work.  I've only done a high level code read and quick test.  I'll test some more and deep dive the code tomorrow morning.

File: assets/component-facets.css
Comment (patrickracicot) [inline assets/component-facets.css:333]: (question) Couldn't we just put the rule as one of the rules in the `:where` since where is a [forgiving selector parsing](https://developer.mozilla.org/en-US/docs/Web/CSS/:where#forgiving_selector_parsing)

File: assets/component-facets.css
Comment (patrickracicot) [inline assets/component-facets.css:303]: (remark) Although this is fine IMO, I just wanted to point out that before, the hover state on the `drawer`/`mobile` mode wouldn't apply and underline to the filter values.  Now this new selectors does have the side-effect of adding an underline style to thoses displays as well.

I do think this is a good change thought.

File: snippets/visual-display.liquid
Comment (patrickracicot) [inline snippets/visual-display.liquid:44]: (question) Any reason you need to return this empty div ?

File: snippets/facets.liquid
Comment (jas7457) [reply snippets/facets.liquid:189]: Weird that only happens in voiceover. It's as if it creates its own border around the element, which is likely what it's doing. To get around this, I just decided to match the height/width of the container. The actual element would go outside of this focus area because of using outline, but this is supposed to be for non-sighted uses anyways so I can't really think of a better (non-hacky) way.

image

File: assets/component-facets.css
Comment (jas7457) [reply assets/component-facets.css:303]: Yeah the non-mobile didn't have an underline before either. Al wanted to add it, he didn't specify if it should only be applied to desktop, but I don't think we should have different styles between the two (or else we'll continue to have spaghetti code in the future)

**Comment by jas7457** [comment]
> (concern) I feel like the active/selected style is not strong enough for thumbnail displays https://screenshot.click/06-48-0c026-zjx41.png

I had a similar concern and brought it to Al and we updated it to what we have now. @al-mcquiston can you weigh in if we should make further adjustments?

File: assets/component-visual-display.css
Comment (dan-menard) [inline assets/component-visual-display.css:93]: Why do you need `!important` here?

File: snippets/facets.liquid
Comment (dan-menard) [inline snippets/facets.liquid:181]: I think you can just do:

```liquid
assign is_disabled = value.count == 0 and value.active == false
```

(and if so, same thing down on line ~715)

File: snippets/visual-display.liquid
Comment (dan-menard) [inline snippets/visual-display.liquid:32]: Love how clear this is 👍

**Review by dan-menard — COMMENTED** [review]
Code and 🎩 look pretty good to me.

@sofiamatulis can you take a look as well, and let me know what I missed? 😅

File: snippets/facets.liquid
Comment (dan-menard) [reply snippets/facets.liquid:181]: Ahh right, liquid doesn't support assigning via ternaries, so I guess that extends to assigning _any_ expression, including a boolean one.

Sorry, my liquid is pretty rusty 🙈

File: snippets/facets.liquid
Comment (andy-schneidr) [inline snippets/facets.liquid:126]: (I think the right place to put this comment is around line 145 but I can't reach ✋)
It could be just something wrong with my store, but when 🎩 ing, the filter `selected` count doesn't seem to work?
image
It was working on Dawn-11.0.0

File: snippets/facets.liquid
Comment (andy-schneidr) [inline snippets/facets.liquid:238]: (question) what does the `-` in `- value.label` do?

**Review by andy-schneidr — COMMENTED** [review]
Did some 🎩  and looked through. Just some comments from me, this is looking really cool! 😄 
A metaobject entry missing text or color/color list/image seems to be handled well!

File: assets/component-facets.css
Comment (jas7457) [reply assets/component-facets.css:825]: Oh I actually forgot that I got rid of highlight completely from the markup for the grid, but what it _was_doing before is it was an invisible element that had `gap` applied because it's now in a display flex with gap. Like I said I got rid of it from my markup later, but I still think it's beneficial to do this because we never want it to take up space, which is what this was doing before.

File: snippets/facets.liquid
Comment (jas7457) [reply snippets/facets.liquid:126]: Yeah, blame @patrickracicot :D https://github.com/Shopify/dawn/pull/3048 - it's fixed in main, I'll rebase and squash in a bit.

File: snippets/facets.liquid
Comment (jas7457) [reply snippets/facets.liquid:238]: It removes whitespace. Adding the class had the formatter push it to the next line, and it actually automatically added the `-` so no new white space was introduced.

Comment (andy-schneidr) [reply snippets/facets.liquid:238]: Ohh, thanks! 👍

File: snippets/facets.liquid
Comment (andy-schneidr) [reply snippets/facets.liquid:126]: Ohh, missed that whole situation, I'm glad it wasn't Just Me 😆

File: snippets/facets.liquid
Comment (NathanPJF) [inline snippets/facets.liquid:119]:
```suggestion
                    assign has_visual_display = true
```

File: snippets/facets.liquid
Comment (NathanPJF) [inline snippets/facets.liquid:122]:
```suggestion
                    assign has_visual_display = false
```

File: snippets/facets.liquid
Comment (NathanPJF) [inline snippets/facets.liquid:201]:
```suggestion
                                {% if has_visual_display %}
```

File: snippets/facets.liquid
Comment (NathanPJF) [inline snippets/facets.liquid:678]:
```suggestion
                        assign has_visual_display = true
```

File: snippets/facets.liquid
Comment (NathanPJF) [inline snippets/facets.liquid:681]:
```suggestion
                        assign has_visual_display = false
```

File: snippets/facets.liquid
Comment (NathanPJF) [inline snippets/facets.liquid:724]:
```suggestion
                                  class="facets__label mobile-facets__label{% if is_disabled %} mobile-facets__label--disabled disabled{% endif %}{% if has_visual_display %} visual-display-parent visual-display-parent--{{ filter.display_style }}{% endif %}{% if value.active %} active{% endif %}"
```

File: snippets/facets.liquid
Comment (NathanPJF) [inline snippets/facets.liquid:740]:
```suggestion
                                  {% if has_visual_display %}
```

File: snippets/facets.liquid
Comment (NathanPJF) [inline snippets/facets.liquid:186]:
```suggestion
                                class="facets__label facet-checkbox{% if is_disabled %} facet-checkbox--disabled disabled{% endif %}{% if has_visual_display %} visual-display-parent visual-display-parent--{{ filter.display_style }}{% endif %}{% if value.active %} active{% endif %}"
```

File: snippets/visual-display.liquid
Comment (NathanPJF) [inline snippets/visual-display.liquid:40]:
```suggestion
      {{ value | image_url: width: 300 | image_tag: class: 'visual-display__image', alt: value.alt }}
```
From our sync: The display size is ~`90`, but in order for this not to look blurry on devices with higher pixel density, (like 3.0x), you'll want to request a higher res image.  

The [Liquid `srcset` attribute](https://shopify.dev/docs/api/liquid/filters/image_tag#image_tag-srcset) doesn't allow us to specify an output like `srcset="imageurl-for-100px 1x, imageurl-for-200px 2x"`, so I recommend hardcoding `300` over `100`.

File: snippets/visual-display.liquid
Comment (NathanPJF) [reply snippets/visual-display.liquid:42]: Idea from our sync: Could use a different element like `span` if you feel it gives slightly less weird CSS declarations on your end. Only reason to do that is to avoid a CSS selector from the screenshot above.

Note: We looked into `figure` + `figurecaption` instead, but that isn't the correct usage of the HTML attribute; especially in the scenario where there's a missing image.

File: assets/component-visual-display.css
Comment (NathanPJF) [inline assets/component-visual-display.css:100]:
```suggestion
.visual-display__image--swatch {
  object-fit: cover;
}

.visual-display__image--thumbnail {
  object-fit: contain;
}
```

^ These classes don't exist yet.


From our sync: **Thumbnail** Images that are much taller than wide (or vice versa) will get a zoomed in look that may make it difficult to tell what the image is. Screenshot example below is showing a Fridge.  I think `contain` is better for thumbnails, even if it results in some filter images having different dimensions. However, when the images are used in a **swatch**, I agree that `cover` is the right implementation


Screenshot 2023-10-06 at 5 39 48 PM


Screenshot 2023-10-06 at 5 39 33 PM
different proportions - the merchant uploaded them that way.

**Review by NathanPJF — CHANGES_REQUESTED** [review]
Summary of requests:
- Consistently calling it `display` over `representation`
- Image size requested from CDN to consider devices with higher pixel density
- Always show full thumbnail image when display type is `thumbnail`

File: assets/component-visual-display.css
Comment (jas7457) [reply assets/component-visual-display.css:100]: I already have top-level classes signifying the style, so I can simply do:
```css
.visual-display--style-swatch .visual-display__image {
  object-fit: cover;
}

.visual-display--style-thumbnail .visual-display__image {
  object-fit: contain;
}
```

File: snippets/visual-display.liquid
Comment (NathanPJF) [inline snippets/visual-display.liquid:23]:
```suggestion
        assign size_limit = value.size | at_most: 4
```

Galaxy brain moment: https://shopify.dev/docs/api/liquid/filters/at_most

**Comment by al-mcquiston** [comment]
> > (concern) I feel like the active/selected style is not strong enough for thumbnail displays https://screenshot.click/06-48-0c026-zjx41.png
> 
> I had a similar concern and brought it to Al and we updated it to what we have now. @al-mcquiston can you weigh in if we should make further adjustments?

Weird I'm seeing something slightly  different (I think) where the contrast is a little more clear IMO. Are you still seeing the same thing @patrickracicot I know we had made some tweaks to things recently.
image

**Review by carolineschnapp — APPROVED** [review]

**Review by patrickracicot — APPROVED** [review]
Works as intended !  I do have some concerns about how subtle the selected/active state is, but I won't block the PR on this.

File: assets/component-facets.css
Comment (sofiamatulis) [inline assets/component-facets.css:309]: When showing multiple colors, it's showing as white. Is that expected? https://screenshot.click/12-00-qb4vh-ppeei.mp4

File: assets/component-facets.css
Comment (sofiamatulis) [inline assets/component-facets.css:322]: I wonder if we have a different threshold for colors in terms of show more? For text, 10 makes sense because the list is long, but for color I feel like we could have more rows before showing `show more`?

https://screenshot.click/12-02-15hgf-y47n7.mp4

File: assets/component-facets.css
Comment (sofiamatulis) [inline assets/component-facets.css:810]: This is happening on mobile: https://screenshot.click/12-04-wo88b-bk2tr.mp4

File: assets/component-facets.css
Comment (sofiamatulis) [inline assets/component-facets.css:828]: Is there a reason why certain colors show the opacity different when unavailable while some have the `/`?

https://screenshot.click/12-06-17qnp-wwrsu.png

File: assets/component-visual-display.css
Comment (sofiamatulis) [reply assets/component-visual-display.css:96]: Can we find a different approach instead of using `important`? 👀 

We can add a class and with specificity, itll ensure you dont need `important` and it displays as `block`

https://screenshot.click/12-14-icau3-783u9.png

File: assets/component-visual-display.css
Comment (sofiamatulis) [inline assets/component-visual-display.css:68]: What is `a` and `b` here? 👀

File: snippets/facets.liquid
Comment (sofiamatulis) [inline snippets/facets.liquid:174]: Since this is `media` I wonder if it should follow the global styles

https://screenshot.click/12-18-7qjvv-eo3le.mp4

File: assets/component-facets.css
Comment (sofiamatulis) [reply assets/component-facets.css:294]: I see we have the special case for white, but we should ensure it works for whatever the background color is 👀 
https://screenshot.click/12-20-93gyq-vif08.mp4

**Review by sofiamatulis — COMMENTED** [review]
This looks good 🎉 Added a few comments/questions

File: assets/component-facets.css
Comment (jas7457) [reply assets/component-facets.css:272]: You're looking at the wrong "Color" filter, remember you had two set up? Since you've searched for "bag", that doesn't match the product you attached the visual filter to, so it won't show up.

You could search for "*" instead and you should see it.

File: assets/component-facets.css
Comment (jas7457) [reply assets/component-facets.css:294]: Hmm we don't have any special casing for white, it should work the same no matter the background color. We put a border around it like we can see in your screenshot, and the hover/focus/active states will make it stand out as well. Do you have any suggestions for any changes? I'd think the border around the swatch/thumbnail is all we need. https://screenshot.click/12-58-s8ep1-tb2lw.mp4

File: assets/component-facets.css
Comment (jas7457) [reply assets/component-facets.css:309]: That's because when we set up the filter, we set the override to Image (which doesn't look like you're using).

image

Instead, you can click the pencil/edit icon and change Swatch override to Colors and save.

File: snippets/facets.liquid
Comment (jas7457) [reply snippets/facets.liquid:174]: We talked about it inheriting from global styles, but decided for the first release, to not have it be based on anything except the button radius for thumbnails. @al-mcquiston may be able to better give background, but it was a conscious decision to not attach to global styles.

File: assets/component-facets.css
Comment (jas7457) [reply assets/component-facets.css:810]: Good call out. I had added some styles that are no longer necessary. Instead, I got rid of those styles and set the width to 100% (to match the grid cell) in the case that we're inside of the `.facets-layout-grid`. That way, if it's used by itself it should have the 9rem, otherwise it fills the full width.

Note: the reason this is necessary is our "default" should be 9rem (if this is used outside of filters), but that doesn't actually fill the width for our desktop filters. If we don't force it to 100% there, we will get extra spacing on the left/right, which we don't want, so I can just force it to 100% width.

image

**Review by andy-schneidr — APPROVED** [review]
lgtm, I think it's a little hard for me to [tell which thumbnails I've selected](https://screenshot.click/12-16-umlz2-vucu6.png) (Only thing I can notice is bolded text)
Even if it's not Just Me, I wouldn't hold the review up over it 👍

File: assets/component-facets.css
Comment (sofiamatulis) [reply assets/component-facets.css:294]: What I mean is that the border around the white is dotted while the others arent. That works when the background is white, but not when you change it to a different color 🤔 

@melissaperreault thoughts?

File: assets/component-facets.css
Comment (jas7457) [reply assets/component-facets.css:832]: We use the current foreground color at 40% opacity, along with setting the opacity of the image itself to 40%. Perhaps we could boost the opacity to make it more apparent, but the opacity itself also signifies it's disabled. Any suggestions?

Comment (al-mcquiston) [reply assets/component-facets.css:832]: I was noticing this too and think it's something we should fix. Is there a way we can just not apply the decreased opacity to the /?

File: assets/component-visual-display.css
Comment (sofiamatulis) [reply assets/component-visual-display.css:69]: Thanks for the clarification. Im not sure if we need this here 🤔 Is there a reason why we are using `focus-within`

In Dawn, we have used `focus` and `focus-visible` historically

https://github.com/search?q=repo%3AShopify%2Fdawn+%3Afocus&type=code

Comment (jas7457) [reply assets/component-visual-display.css:69]: So this is a pretty unique scenario, but I essentially want to know if _anything_ inside of `visual-display-parent` has the `:focus-visible` (thus only visible if you use your keyboard to focus, not clicking, and then style the `.visual-display` inside of it. There isn't a css pseudo class that allows this, and there's a pretty good [article](https://larsmagnus.co/blog/focus-visible-within-the-missing-pseudo-class) about it.

Essentially I have this class `.visual-display-parent:has(:focus-visible) .visual-display--style-swatch` giving us the preferred css selector, then this `@supports not` style will be the fallback for browsers that don't yet support `:has`, giving them the less ideal `:focus-within` experience.

Here's a [sandbox](https://codesandbox.io/s/holy-fast-mdnhsq?file=/src/styles.css) and a [video](https://screenshot.click/12-40-j67lj-2tk2h.mp4) showing the difference!

File: assets/component-facets.css
Comment (melissaperreault) [reply assets/component-facets.css:326]: - +1 for using a multiple of 3, it's great to have some rules here to structure the scannability.
- For the limit I'll let @al-mcquiston provide the guidance. 24 seems a lot considering the color labels length and how they wrap on multiple lines.

File: snippets/facets.liquid
Comment (jas7457) [reply snippets/facets.liquid:176]: > Have we explored not inheriting at all until we have a better home for such settings?

I am aligned not doing _any_ global settings for this first iteration. I can remove the border radius.

> I would recommend increasing the bottom padding of those thumbnails to improve the white balance.

Updated to 8px, good call out.

> Have we considered reducing the label size as well to help make more room for more letters?

I updated to 1.3rem instead of 1.4rem. It's not a huge difference, but 1.2 looked way too tiny compared to the other filter text.

> Have we explored making the selected state thicker?

We've heard this feedback a few times. @al-mcquiston do you think we should make any changes since it's pretty common feedback?

**Comment by mkoenig-shopify** [comment]
I know it's a bit late in the game here but I do really wonder if for v1 here we can just ship swatch and text - and then iterate if we truly need thumbnail? i know we could rename it as "croppable" or "not-croppable" but still, feels like we don't know for sure we need that yet and the complexity it adds is high

**Review by NathanPJF — APPROVED** [review]

**Review by sofiamatulis — APPROVED** [review]
Looks good! thanks for addressing the feedback 🎉 Let's just make sure we address these before merging

- https://github.com/Shopify/dawn/pull/3045#discussion_r1358395991
- https://github.com/Shopify/dawn/pull/3045#discussion_r1357329830
- https://github.com/Shopify/dawn/pull/3045#discussion_r1358511327

File: snippets/facets.liquid
Comment (jas7457) [reply snippets/facets.liquid:176]: > We've heard this feedback a few times. @al-mcquiston do you think we should make any changes since it's pretty common feedback?

To follow up on this: I've made some changes that Al and I talked about and this has been pushed. We're chatting about the `/` a bit now and might make a couple slight tweaks to the focus, but then we should be good!

File: assets/component-facets.css
Comment (jas7457) [reply assets/component-facets.css:339]: Changed to use 15 for swatch (5 rows), 12 for thumbnail (4 rows)

**Comment by melissaperreault** [comment]
Thank you so much @jas7457 for all the updates! 🙌 

Capturing notes from our DM conversations here:
- Contrast between the default and selected state looks good! 🎉 💪 
- I didn't get the chance to over stress test the limit before we truncate the thumbnails or swatches to reveal the show more link but I trust this was heavily tested 🙏 

### Not a blocker, but worth investigating
**Hard coded color**: Small hesitation for using a hard coded color for the disabled diagonal line. Why? Let me explain!

_**Summary:**_ Have we explored changing the brightness of the text (foreground color of a scheme) to inform the result over the thumbnails and color swatches? If we observed that text color can't be dark/light enough to be noticeable used at a certain opacity, it may be a safer solution rather than relying on a hard coded color that merchants can't change easily. This way, the result with some opacity may appear as strong as you are expecting it to be initially.

**Potential issues**
- When text color is colorful, that hard coded color can become unexpected. ([Dramatic example](https://screenshot.click/18-34-0rs9h-5a5lk.png))
- When building a theme, when we use a hard coded color we ensure it will be contrasting enough. e.g. The Pickup location checkmark white outline.

**[Debut](https://debut-theme-lightroom.myshopify.com) use case**
Depending on the background color brightness of the primary button we made the hover lighter or darker identifying a specific threshold for contrast.

See other examples: see example 10 and 11 https://www.w3.org/TR/css-color-5/#relative-colors

**Review by NathanPJF — CHANGES_REQUESTED** [review]
Hold off as we discuss API changes

**Review by NathanPJF — APPROVED** [review]
🎩 'd and new API useage 👍🏻

**Pull request merged**
